BGDIINF_SB-1264: combined new item and collection serializers

hansmannj opened the pull request:
Serializers for collection and items should now work.
Collection serializer (with JSONFields for summaries and extent) in both ways, items only serialization so far.
Asset model is adapted to be compatible with new collections model.

ltshb commented on app/stac_api/models.py:
```diff
@@ -120,46 +118,44 @@ class Collection(models.Model):
-    summaries_eo_gsd = ArrayField(models.FloatField(), blank=True, null=True)
-    summaries_proj = ArrayField(models.IntegerField(), blank=True, null=True)
+    summaries = models.JSONField(blank=True, null=True)
+    #summaries_eo_gsd = ArrayField(models.FloatField(), blank=True, null=True)
+    #summaries_proj = ArrayField(models.IntegerField(), blank=True, null=True)
```
These should be removed

hansmannj replied:
done

ltshb commented on app/stac_api/models.py:
```diff
@@ -120,46 +118,44 @@ class Collection(models.Model):
     # update of an asset
-    geoadmin_variant = ArrayField(models.CharField(max_length=15), blank=True, null=True)
-
+    #geoadmin_variant = ArrayField(models.CharField(max_length=15), blank=True, null=True)
```
Clean up

hansmannj replied:
as for all of your "KeyError" comments: I will add handling, but as these fields are only for internal write access, the requested keys and indices should always work. But yes, if something is wrong with the internal writing, this would cause errors. I'll add that soon

ltshb commented on app/stac_api/models.py:
```diff
@@ -120,46 +118,44 @@ class Collection(models.Model):
     def clean(self):
         # TODO: move this check to the items save()
-        if self.start_date is None and self.end_date is None:
+        if self.extent["temporal"]["interval"][0][0] is None and self.extent["temporal"][
```
Exception handling is missing; KeyError and IndexError must be handle here

hansmannj replied:
added exception handling here and also (hopefully) everywhere, where Key- or IndexErrors might occur

ltshb commented on app/stac_api/models.py:
```diff
@@ -120,46 +118,44 @@ class Collection(models.Model):
         # very simple validation, raises error when geoadmin_variant strings contain special
         # characters or umlaut.
-        for variant in self.geoadmin_variant:
+        for variant in self.summaries["geoadmin:variant"]:
```
KeyError must be handle here otherwise it will crash if the `geoadmin:variant` is missing

ltshb commented on app/stac_api/models.py:
```diff
@@ -210,17 +206,20 @@ def save(self, force_insert=False, force_update=False, using=None, update_fields
         # check if collection's start_ and end_dates need to be updated
-        if self.collection.start_date is None:
-            self.collection.start_date = self.properties_datetime
+        if self.collection.extent["temporal"]["interval"][0][0] is None:
```
Exception of `KeyError` and `IndexError`

ltshb commented on app/stac_api/models.py:
```diff
@@ -210,17 +206,20 @@ def save(self, force_insert=False, force_update=False, using=None, update_fields
-            self.collection.start_date = self.properties_datetime
+        if self.collection.extent["temporal"]["interval"][0][0] is None:
+
+            self.collection.extent["temporal"]["interval"][0][0] = self.properties_datetime
```
Exception of `KeyError` and `IndexError`

ltshb commented on app/stac_api/models.py:
```diff
@@ -210,17 +206,20 @@ def save(self, force_insert=False, force_update=False, using=None, update_fields
 
-        elif self.properties_datetime < self.collection.start_date:
-            self.collection.start_date = self.properties_datetime
+        elif self.properties_datetime < self.collection.extent["temporal"]["interval"][0][0]:
```
Exception of `KeyError` and `IndexError`

ltshb commented on app/stac_api/models.py:
```diff
@@ -210,17 +206,20 @@ def save(self, force_insert=False, force_update=False, using=None, update_fields
-            self.collection.start_date = self.properties_datetime
+        elif self.properties_datetime < self.collection.extent["temporal"]["interval"][0][0]:
+
+            self.collection.extent["temporal"]["interval"][0][0] = self.properties_datetime
```
Exception of `KeyError` and `IndexError`

ltshb commented on app/stac_api/models.py:
```diff
@@ -210,17 +206,20 @@ def save(self, force_insert=False, force_update=False, using=None, update_fields
-        if self.collection.end_date is None \
-            or self.properties_datetime > self.collection.end_date:
-            self.collection.end_date = self.properties_datetime
+        elif self.properties_datetime > self.collection.extent["temporal"]["interval"][0][1] or \
```
Exception of `KeyError` and `IndexError`

ltshb commented on app/stac_api/models.py:
```diff
@@ -281,24 +280,24 @@ def save(self, force_insert=False, force_update=False, using=None, update_fields
         # check if the collection's geoadmin_variant needs to be updated
-        if self.geoadmin_variant not in self.feature.collection.geoadmin_variant:
-            self.feature.collection.geoadmin_variant.append(self.geoadmin_variant)
+        if self.geoadmin_variant not in self.feature.collection.summaries["geoadmin:vairant"]:
```
Exception of `KeyError`

ltshb commented on app/stac_api/models.py:
```diff
@@ -281,24 +280,24 @@ def save(self, force_insert=False, force_update=False, using=None, update_fields
         # and grandparent level.
-        if self.proj_epsq not in self.feature.collection.summaries_proj:
-            self.feature.collection.summaries_proj.append(self.proj_epsq)
+        if self.proj_epsq not in self.feature.collection.summaries["proj:epsg"]:
```
Exception of `KeyError`

ltshb commented on app/stac_api/models.py:
```diff
@@ -281,24 +280,24 @@ def save(self, force_insert=False, force_update=False, using=None, update_fields
 
-        if not float_in(self.eo_gsd, self.feature.collection.summaries_eo_gsd):
-            self.feature.collection.summaries_eo_gsd.append(self.eo_gsd)
+        if not float_in(self.eo_gsd, self.feature.collection.summaries["es:gsd"]):
```
Exception of `KeyError`

ltshb reviewed (changes requested):
Error handling for the Json field is missing, it needs to handle some cases otherwise the application will crash.

boecklic commented on app/stac_api/models.py:
```diff
@@ -125,48 +137,52 @@ class Collection(models.Model):
+        try:
+            for variant in self.summaries["geoadmin:variant"]:
+                if not bool(re.search('^[a-zA-Z0-9]*$', variant)):
+                    raise ValidationError(_('Property geoadmin:variant not correctly specified.'))
```
The error message could also give hints on what was wrong (we don't have security implications here, so we can very well tell the user what exactly was wrong or what we expect)

hansmannj replied:
error messages are now more verbose

boecklic commented on app/stac_api/models.py:
```diff
@@ -214,19 +230,35 @@ def save(self, force_insert=False, force_update=False, using=None, update_fields
-        elif self.properties_datetime < self.collection.start_date:
-            self.collection.start_date = self.properties_datetime
-            self.collection.save()
+                self.collection.extent["temporal"]["interval"][0][0] = self.properties_datetime
```
`extent["temporal"]["interval"][0][0]` <- this smells a bit like it could cause troubles if for some reason the extent field doesn't have this exact structure. This should never happen, but still...

hansmannj replied:
Every access to the JSON-fields reps. dictionaries should now be inside a try ... except statement, so that at least an error message will be raised and logged, in case something unexpected happens

boecklic commented on app/stac_api/models.py:
```diff
@@ -277,37 +309,66 @@ def __str__(self):
-            self.feature.collection.save()
+        try:
+            # check if the collection's geoadmin_variant needs to be updated
+            if self.feature.collection.summaries["geoadmin:variant"] is None:
+
+                self.feature.collection.summaries["geoadmin:variant"] = [self.geoadmin_variant]
+                self.feature.collection.save()
```
For these checks it would maybe be cleaner to create little helpers/wrappers in the collection model, like `update_variants` or the like. This would keep the internal logic of these fields encapsulated to the collection, item in the end doesn't need to know about how these fields are organized internally. If we change it, it's 'just' a change in the internal representation in the collection and doesn't need to interest others (in the spirit of OOP)

hansmannj replied:
Good point, done!

ltshb commented on app/stac_api/models.py:
```diff
@@ -125,48 +145,137 @@ class Collection(models.Model):
+        updated or an error will be raised, if updating fails.
+        '''
+        try:
+            if self.summaries["geoadmin:variant"] is None:
+
+                self.summaries["geoadmin:variant"] = [asset_geoadmin_variant]
+                self.save()
+
+            elif asset_geoadmin_variant not in self.summaries["geoadmin:variant"]:
+                self.summaries["geoadmin:variant"].append(asset_geoadmin_variant)
+                self.save()
 
+            if self.summaries["proj:epsg"] is None:
+                self.summaries["proj:epsg"] = [asset_proj_epsg]
+                self.save()
+
+            elif asset_proj_epsg not in self.summaries["proj:epsg"]:
+                self.summaries["proj:epsg"].append(asset_proj_epsg)
+                self.save()
+
+            if self.summaries["eo:gsd"] is None:
+                self.summaries["eo:gsd"] = [asset_eo_gsd]
+                self.save()
+
+            elif not float_in(asset_eo_gsd, self.summaries["eo:gsd"]):
+                self.summaries["eo:gsd"].append(asset_eo_gsd)
+                self.save()
```
Just a minor code style, I would group the `if ... elif` without blank lines and have only blank line between `if ... elif` block this make it easier to read.

hansmannj replied:
done, I'll check, if that might be a bug in yapf or was a bug in my brain ;-)

ltshb commented on app/stac_api/models.py:
```diff
@@ -125,48 +145,137 @@ class Collection(models.Model):
+                self.extent["temporal"]["interval"][0][0] = item_properties_datetime
+                self.save()
+
+            elif self.extent["temporal"]["interval"][0][1] is None:
```
Should not it be `if` here ? and same comment about blank lines as above.

hansmannj replied:
From my understanding, `elif` is correct, but the whole block is quite nasty to read.

When reaching the line (formerly known as line 245), so far it has been checked, that:
* start_date in the extent's temporal inteval is **not** `None` **and**
* current item instance's datetime value is **not** < extent's temporal interval's start_date, so **only** then:
* check if end_date in extent's temporal interval is `None` and if so,
* set that to the current item's instances datetime value

Hence, it should be an `eflif` and not only an `if` here, from my understanding.

ltshb commented on app/stac_api/models.py:
```diff
@@ -125,48 +145,137 @@ class Collection(models.Model):
+                self.save()
+
+        except (KeyError, IndexError) as err:
+            logger.error(_(
```
I would not use `_()` getext in logging. Getext is for localization and we don't want to localize logging. All logging must be written in English. Localization should be kept for user interface message (e.g. like in the `ValidationError` where the message is for the user)

hansmannj replied:
make sense, I've changed that

ltshb commented on app/stac_api/models.py:
```diff
@@ -125,48 +145,137 @@ class Collection(models.Model):
+
+        except (KeyError, IndexError) as err:
+
+            logger.error(_('Updating the collection extent due to item update failed: %s'), err)
```
I would not use `_()` getext in logging. Getext is for localization and we don't want to localize logging. All logging must be written in English. Localization should be kept for user interface message (e.g. like in the `ValidationError` where the message is for the user)

ltshb commented on app/stac_api/models.py:
```diff
@@ -125,48 +145,137 @@ class Collection(models.Model):
+        try:
+            for variant in self.summaries["geoadmin:variant"]:
+                if not bool(re.search('^[a-zA-Z0-9]*$', variant)):
+                    raise ValidationError(_("Property geoadmin:variant not compatible with the" +
```
A log message would be appropriate here as well.

hansmannj replied:
yeah, why not :-)

ltshb commented on app/stac_api/models.py:
```diff
@@ -125,48 +145,137 @@ class Collection(models.Model):
+        except (KeyError, IndexError) as err:
+            # we should only land here, if the default values for the extent are corrupted.
+            # they can only be overriden by internal write processes.
+            logger.error(_("Error when trying to access property geoadmin:variant from inside " +
```
I would not use `_()` getext in logging. Getext is for localization and we don't want to localize logging. All logging must be written in English. Localization should be kept for user interface message (e.g. like in the `ValidationError` where the message is for the user)

ltshb commented on app/stac_api/models.py:
```diff
@@ -209,24 +318,40 @@ class Item(models.Model):
+            if self.properties_eo_gsd is None:
+                self.properties_eo_gsd = [asset_eo_gsd]
+                self.save()
+
```
Remove blank line

ltshb commented on app/stac_api/models.py:
```diff
@@ -209,24 +318,40 @@ class Item(models.Model):
+                self.save()
+
+        except (KeyError, IndexError) as err:
+            logger.error(_(
```
Remove getext

ltshb commented on app/stac_api/models.py:
```diff
@@ -277,37 +402,29 @@ def __str__(self):
         # characters or umlaut.
-        if not bool(re.search('^[a-zA-Z0-9]*$', self.geoadmin_variant)):
-            raise ValidationError(_('Property geoadmin:variant not correctly specified.'))
+        try:
+            if not bool(re.search('^[a-zA-Z0-9]*$', self.geoadmin_variant)):
+                raise ValidationError(_("Property geoadmin:variant not compatible with the " +
+                                        "naming conventions."))
+
+        except (KeyError, IndexError) as err:
+            logger.error(_(
+                "Error when trying to access property geoadmin:variant from within asset's " +
+                "clean function: %s"),
+                         err
+                        )
+            raise ValidationError(_(
+                "Error when trying to access property geoamin:variant from within asset's " +
+                "clean function"
+            ))
```
Remove the `try ... except` which is not needed here. And add a log in the case of validation error

hansmannj replied:
Done

ltshb reviewed (changes requested):
Good job, only issues with the `if ... elif` block and the removing the `_()` getext from logging. We need to synchronize to see who is merging first with PR #56 and #57.

What is the purpose of the `scripst/test_serializer.py` ? should this one not be removed and only use the unit test ?

boecklic reviewed (dismissed):
nice!
One little thing: `scripts/test_serializer.py` should be transformed to a regular unit test

hansmannj commented:
> nice!
> One little thing: `scripts/test_serializer.py` should be transformed to a regular unit test

I added extra unit tests. The "test_serializer.py" had a wrong name. It was a script to fill the local test db with some values (until the manage.py fill_db_command works). I renamed it, to avoid confusions.

ltshb commented on app/stac_api/models.py:
```diff
@@ -209,24 +314,39 @@ class Item(models.Model):
+                self.properties_eo_gsd.append(asset_eo_gsd)
+                self.save()
+
+        except (KeyError, IndexError) as err:
```
Remove unnecessary `try ... except`

hansmannj replied:
done

ltshb commented on app/stac_api/models.py:
```diff
@@ -125,48 +145,131 @@ class Collection(models.Model):
+                    raise ValidationError(_("Property geoadmin:variant not compatible with the"
+                                            "naming conventions."))
+
+        except (KeyError, IndexError) as err:
```
Remove unnecessary `IndexError` from `try .... except`

hansmannj replied:
done

ltshb commented on app/stac_api/models.py:
```diff
@@ -125,48 +145,131 @@ class Collection(models.Model):
+                self.summaries["eo:gsd"].append(asset_eo_gsd)
+                self.save()
+
+        except (KeyError, IndexError) as err:
```
Remove unnecessary `IndexError` from `try .... except`

hansmannj replied:
done

hansmannj replied:
this one should stay I guess

ltshb replied:
It should be `except KeyError as err:` as there should not have any `IndexError` in this code, only `KeyError`

ltshb reviewed (approved):
Looks great :+1: Just minor clean up in the `try ... except` error handling, see inline comments
